modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/java/property/wrappertype/WrapperTypeUtils.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.loader.java.property.wrappertype;
+
+import static java.lang.String.format;
+import org.mule.runtime.api.meta.model.ModelProperty;
+import org.mule.runtime.module.extension.internal.runtime.resolver.ValueResolver;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * Utility class to reuse logic around {@link WrapperType}
+ *
+ * @since 4.0
+ */
+public class WrapperTypeUtils {
+
+  /**
+   * Given a expression value, the expected type and the stack of {@link WrapperType wrapper types}, iterates it and
+   * creates all the required {@link ValueResolver} and stacks them is necessary.
+   *
+   * @param expression           Expression value for the expression based value resolver.
+   * @param expectedType         The expected type of the expression resolution
+   * @param wrapperModelProperty Model property containing the stack of {@link WrapperType wrapper types}
+   * @return The expression based {@link ValueResolver}
+   */
+  public static ValueResolver getExpressionBasedWrapperValueResolver(String expression, Class expectedType,
```

Comment:
most of this methods are the sort of thing that I would expect as instance methods of a ResolverStack class. Hopefully, such a change makes refactoring easier.